Sending focus to modal in CMS

clrux opened the pull request:
# Modal focus management

The work here focuses on focus management for CMS modal windows, specifically Xblocks when editing added components.

## History

Previously, when editing a component, focus was not sent to the modal and so editing the component was not possible. This fix addresses that accessibility issue by doing the following:

* adds `tabindex="-1"` to the modal containers so they are able to receive focus, 
* adds `aria-labelledby` to the modal container, identified by the title IDs so context is provided, and
* sends `.focus()` to the modal after it is rendered and resized

---

## Browser testing

Ensures the functionality here works in our browsers.

| Platform | Browser | Version | Passed |
|----------|---------|---------|--------|
| Mac      | Chrome  | 43      | Yes    |
|          | Safari  | 8       | Yes    |
|          | Firefox | 24      | Yes    |
| Windows  | Chrome  | 43      | Yes    |
|          | Firefox | 38      | Yes    |
|          | IE      | 10      | Yes    |
|          | IE      | 11      | Yes    |

--- 

## Issues remaining not covered in this piece of work

Even though focus is now being sent to the modal for further interaction, other expected keyboard functionality is still missing. We may want to consider revisiting these items in future piece of work:

* pressing `Esc` doesn't close the modal
* closing the modal doesn't send focus back to the original initiating element
* in modals that contain a text editor, focus goes there first making this somewhat inconsistent
* it is possible to tab beyond the modal, going back to the base DOM; from there it nearly impossible to get back into the open modal

---

### Reviewers
- [x] @cahrens (Python/JS) 
- [ ] @cptvitamin (Accessibility)

clrux commented:
jenkins run js
jenkins run bokchoy
jenkins run python

cahrens commented on cms/static/js/views/modals/base_modal.js:
```diff
@@ -87,6 +87,9 @@ define(["jquery", "underscore", "gettext", "js/views/baseview"],
                 $(window).resize(_.bind(this.resize, this));
+
+                // after showing and resizing, send focus
+                this.$el.find('.modal-window').focus();
```
The class ".modal-window" is used 2 times in this file. Please promote it to a class variable to make the code easier to maintain.

cahrens commented:
Please add a test case in /edx-platform/cms/static/js/spec/views/modals/base_modal_spec.js (or extend an existing test case, if appropriate) to cover this code change. Make sure that the test fails before your fix.

clrux commented:
Thanks @cahrens. I've added the necessary test to check that focus was sent.

cahrens commented on cms/static/js/views/modals/base_modal.js:
```diff
@@ -34,6 +34,7 @@ define(["jquery", "underscore", "gettext", "js/views/baseview"],
                 modalType: 'generic',
                 modalSize: 'lg',
                 title: '',
+                modalFocusWindow: '.modal-window',
```
A more descriptive name would be modalWindowClass (it has nothing to do with focus per se).

cahrens commented on cms/static/js/spec_helpers/modal_helpers.js:
```diff
@@ -32,6 +32,13 @@ define(["jquery", "common/js/spec_helpers/template_helpers", "js/spec_helpers/vi
+            var modalElement = getModalElement(modal),
+                modalFocusWindow = modalElement.find('.modal-window');
+
+            return modalFocusWindow.focus();
```
As we discussed, you don't want to actually call focus in the test method-- you want to verify if focus was called on the modal as a result of show being called.

clrux commented:
@cahrens I've done a few things:

* simplified the `modal_helpers.js` line to return if the modal has focus
* cleaned up the test itself to check only for true
* created a reusable class in `base_modal.js` for the modal window classname

Oddly though, focus is sent to the modal correctly and consistently when testing this in the CMS. The tests, however, will sometimes randomly fail. Could this be a race condition?

clrux commented:
jenkins test python

clrux commented:
jenkins run python

cahrens commented on cms/static/js/views/modals/base_modal.js:
```diff
@@ -146,7 +150,7 @@ define(["jquery", "underscore", "gettext", "js/views/baseview"],
                     availableWidth, availableHeight, maxWidth, maxHeight;
 
-                modalWindow = this.$('.modal-window');
+                modalWindow = this.$el.find(this.options.modalWindowClass)
```
Missing semi-colon at the end. Make sure to enable JSHint or JSLint in your editor.

cahrens commented:
Note that focus does not go into the settings editor when you edit a discussion module. That component has no "text tab", only the settings tab.

cahrens commented on cms/static/js/views/modals/base_modal.js:
```diff
@@ -34,6 +34,7 @@ define(["jquery", "underscore", "gettext", "js/views/baseview"],
                 modalType: 'generic',
                 modalSize: 'lg',
                 title: '',
+                modalWindowClass: '.modal-window',
```
Please replace the other places in this file that specify ".modal-window" with your new variable.

cahrens commented on cms/static/js/spec/views/modals/base_modal_spec.js:
```diff
@@ -31,6 +31,12 @@ define(["jquery", "underscore", "js/views/modals/base_modal", "js/spec_helpers/m
+                it('sends focus to the modal window after show is called', function() {
+                    showMockModal();
+                    expect(ModelHelpers.getModalWindow(modal)).toBeFocused();
+                    // expect(ModelHelpers.isModalFocused(modal)).toBeTruthy();
```
Delete this comment.

clrux commented:
@cahrens Thanks for all of the feedback. I updated the description to cover the additional cases found where focus doesn't function as intended and the note about going beyond the modal, for future work. This PR is meant to be scoped to actually giving focus to the modal.

I made a few changes which allow my local tests (via `test_js_dev -s cms` on Chrome) to pass consistently. However, when I run `test_js -s cms` or `paver test_js_run -s cms` (which uses Firefox) the tests consistently fail. Why would the tests work when running the dev set of tests, but fail otherwise? Is this possibly related to the order the tests run?

I'm looping in @frrrances here because testing really isn't my area of expertise. As much as I want to learn it, I can't continue spending a full day trying to figure out something like this, when an engineer who's more familiar could probably knock this out in an hour or two. I think I may need some engineering help to get this over the line. What is the best way to get this prioritized?

cahrens commented on cms/static/js/views/modals/base_modal.js:
```diff
@@ -146,7 +151,7 @@ define(["jquery", "underscore", "gettext", "js/views/baseview"],
                 var top, left, modalWindow, modalWidth, modalHeight,
                     availableWidth, availableHeight, maxWidth, maxHeight;
 
-                modalWindow = this.$('.modal-window');
```
Still missing a semi-colon at the end of line 154.

cahrens commented:
@clrux I pushed up a change that passes for me locally. While searching for other usages of "toBeFocused", I saw that many of them had been skipped because the tests ended up being flaky. I'm hoping that using a wait will allow this test to be stable. It is ugly that I copied the guts of toBeFocused, but unfortunately that big of logic exists in many places, without a clean way to reuse it. I would certainly be supportive of creating a helper method.

Let's see how this does in Jenkins, and if all looks good, I am :+1: once you fix the minor JS issue I noted above.

cahrens commented:
Make sure to squash commits before you merge.
